### PR TITLE
Generate: `can_generate()` recursive check

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

The deprecation warning added in #33203 was incomplete: classes inheriting from classes that have received the update were throwing the warning. This PR adds a recursive check to `can_generate()` and tests the warnings thrown by the function.

Example of script that is throwing a warning, but shouldn't (and is fixed in this PR):
```py
from transformers.models.llama import LlamaForCausalLM

class NewLlamaForCausalLM(LlamaForCausalLM):
    pass

model = NewLlamaForCausalLM.from_pretrained("hf-internal-testing/tiny-random-LlamaForCausalLM")
model.generate()
```

Thank you @regisss for warning me about this one and providing a reproducer 🙏

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
Ok! @regisss can you comment whether this fixes the issue you encountered?

**Comment by regisss** [comment]
> Ok! @regisss can you comment whether this fixes the issue you encountered?

Yep it's all good on my side, the warning doesn't appear anymore with this fix. Thanks!

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -1645,6 +1645,12 @@ def can_generate(cls) -> bool:
         # Model class overwrites `generate` (e.g. time series models) -> can generate
         if str(cls.__name__) in str(cls.generate):
             return True
+        # The class inherits from a class that can generate (recursive check) -> can generate
+        for base in cls.__bases__:
+            if not hasattr(base, "can_generate"):
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:1650]: is this only grandfather check?

**Review by ArthurZucker — APPROVED** [review]
Thanks, wil have the same issue for great grand pa check no?

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (gante) [reply src/transformers/modeling_utils.py:1650]: it's recursive! (it calls `can_generate` of the parent, which in turn may call the granparent's, and so on :D )

**Pull request merged**